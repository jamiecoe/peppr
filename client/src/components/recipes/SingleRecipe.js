import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import shortid from 'shortid';
import { getSingleRecipe, deleteRecipe } from '../../actions/recipes';
import Navbar from '../Navbar';

export class SingleRecipe extends Component {
  componentDidMount() {
    if (!this.props.recipe) {
      const { id } = this.props.match.params;
      this.props.getSingleRecipe(id);
    }
  }

  onDeleteClick = () => {
    const { id } = this.props.match.params;
    this.props.deleteRecipe(id);
  };

  renderStringToList(str) {
    const stringSplit = str.includes('\n\n') ? '\n\n' : '\n';
    return str.split(stringSplit).map((line, index) => {
      if (index !== str.split(stringSplit).length - 1) {
        return <li key={shortid.generate()}>{line}</li>;
      }
      return null;
    });
  }

  render() {
    const { recipe } = this.props;

    if (!recipe) return <div>Loading...</div>;

    return (
      <div className="singleRecipe__container">
        <div
          className="singleRecipe__img"
          style={{ backgroundImage: `url(${recipe.imageurl})` }}
        />
        <div className="singleRecipe__textcontainer--outer">
          <div className="singleRecipe__textcontainer--inner">
            <Link className="singleRecipe__backarrow" to="/recipes">
              <i className="ion-ios-arrow-back" />
            </Link>
            <h1 className="singleRecipe__title">{recipe.title}</h1>
            <h2 className="singleRecipe__sub-title">Ingredients</h2>
            <ul className="singleRecipe__list--ingredients">
              {this.renderStringToList(recipe.ingredients)}
            </ul>
            <h2 className="singleRecipe__sub-title">Method</h2>
            <ol className="singleRecipe__list--method">
              {this.renderStringToList(recipe.method)}
            </ol>
            <button
              className="singleRecipe__deleteButton"
              onClick={this.onDeleteClick}
            >
              Delete Recipe
            </button>
          </div>
        </div>
        <Navbar page="singleRecipe" />
      </div>
    );
  }
}

const mapStateToProps = ({ recipes }, ownProps) =>
  (recipes ? { recipe: recipes[ownProps.match.params.id] } : { recipe: recipes });

export default connect(mapStateToProps, { getSingleRecipe, deleteRecipe })(SingleRecipe);
