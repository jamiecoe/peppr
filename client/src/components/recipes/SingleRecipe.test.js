import React from 'react';
import { shallow } from 'enzyme';
import { SingleRecipe } from './SingleRecipe';

describe('SingleRecipe', () => {
  const mockRecipe = {
    imageurl: '',
    ingredients: '',
    method: '',
    title: '',
  };

  const mockGetSingleRecipe = jest.fn();
  const mockDeleteRecipe = jest.fn();

  const props = {
    recipe: mockRecipe,
    getSingleRecipe: mockGetSingleRecipe,
    deleteRecipe: mockDeleteRecipe,
    match: {
      params: {
        id: 2,
      },
    },
  };

  let singleRecipe = shallow(<SingleRecipe {...props} />);

  it('renders properly', () => {
    expect(singleRecipe).toMatchSnapshot();
  });

  describe('when user clicks on delete button', () => {
    beforeEach(() => {
      singleRecipe.find('.singleRecipe__deleteButton').simulate('click');
    });

    it('should call the `deleteRecipe` method it receives from props with the id on the url params', () => {
      expect(mockDeleteRecipe).toHaveBeenCalledWith(props.match.params.id);
    });
  });

  describe('when `recipe` props has not been set yet', () => {
    beforeEach(() => {
      props.recipe = undefined;
      props.getSingleRecipe = mockGetSingleRecipe;
      singleRecipe = shallow(<SingleRecipe {...props} />, {
        disableLifecycleMethods: false,
      });
    });

    it('should only render a loading message <div>', () => {
      expect(singleRecipe).toMatchSnapshot();
    });

    it('calls the `getSingleRecipe` method it receives from props on componentDidMount() with the id on the url params', () => {
      expect(mockGetSingleRecipe).toHaveBeenCalledWith(props.match.params.id);
    });
  });
});
