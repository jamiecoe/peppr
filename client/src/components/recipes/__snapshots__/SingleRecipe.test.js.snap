// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleRecipe renders properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SingleRecipe
    deleteRecipe={[Function]}
    getSingleRecipe={[Function]}
    match={
        Object {
            "params": Object {
              "id": 2,
            },
          }
    }
    recipe={
        Object {
            "imageurl": "",
            "ingredients": "",
            "method": "",
            "title": "",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="singleRecipe__img"
          style={
                    Object {
                              "backgroundImage": "url()",
                            }
          }
/>,
        <div
          className="singleRecipe__textcontainer--outer"
>
          <div
                    className="singleRecipe__textcontainer--inner"
          >
                    <Link
                              className="singleRecipe__backarrow"
                              replace={false}
                              to="/recipes"
                    >
                              <i
                                        className="ion-ios-arrow-back"
                              />
                    </Link>
                    <h1
                              className="singleRecipe__title"
                    />
                    <h2
                              className="singleRecipe__sub-title"
                    >
                              Ingredients
                    </h2>
                    <ul
                              className="singleRecipe__list--ingredients"
                    >
                              
                    </ul>
                    <h2
                              className="singleRecipe__sub-title"
                    >
                              Method
                    </h2>
                    <ol
                              className="singleRecipe__list--method"
                    >
                              
                    </ol>
                    <button
                              className="singleRecipe__deleteButton"
                              onClick={[Function]}
                    >
                              Delete Recipe
                    </button>
          </div>
</div>,
        <Connect(Navbar)
          page="singleRecipe"
/>,
      ],
      "className": "singleRecipe__container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "singleRecipe__img",
          "style": Object {
            "backgroundImage": "url()",
          },
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="singleRecipe__textcontainer--inner"
>
            <Link
                        className="singleRecipe__backarrow"
                        replace={false}
                        to="/recipes"
            >
                        <i
                                    className="ion-ios-arrow-back"
                        />
            </Link>
            <h1
                        className="singleRecipe__title"
            />
            <h2
                        className="singleRecipe__sub-title"
            >
                        Ingredients
            </h2>
            <ul
                        className="singleRecipe__list--ingredients"
            >
                        
            </ul>
            <h2
                        className="singleRecipe__sub-title"
            >
                        Method
            </h2>
            <ol
                        className="singleRecipe__list--method"
            >
                        
            </ol>
            <button
                        className="singleRecipe__deleteButton"
                        onClick={[Function]}
            >
                        Delete Recipe
            </button>
</div>,
          "className": "singleRecipe__textcontainer--outer",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Link
                className="singleRecipe__backarrow"
                replace={false}
                to="/recipes"
>
                <i
                                className="ion-ios-arrow-back"
                />
</Link>,
              <h1
                className="singleRecipe__title"
/>,
              <h2
                className="singleRecipe__sub-title"
>
                Ingredients
</h2>,
              <ul
                className="singleRecipe__list--ingredients"
>
                
</ul>,
              <h2
                className="singleRecipe__sub-title"
>
                Method
</h2>,
              <ol
                className="singleRecipe__list--method"
>
                
</ol>,
              <button
                className="singleRecipe__deleteButton"
                onClick={[Function]}
>
                Delete Recipe
</button>,
            ],
            "className": "singleRecipe__textcontainer--inner",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <i
                  className="ion-ios-arrow-back"
/>,
                "className": "singleRecipe__backarrow",
                "replace": false,
                "to": "/recipes",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "ion-ios-arrow-back",
                },
                "ref": null,
                "rendered": null,
                "type": "i",
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "",
                "className": "singleRecipe__title",
              },
              "ref": null,
              "rendered": "",
              "type": "h1",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Ingredients",
                "className": "singleRecipe__sub-title",
              },
              "ref": null,
              "rendered": "Ingredients",
              "type": "h2",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  null,
                ],
                "className": "singleRecipe__list--ingredients",
              },
              "ref": null,
              "rendered": Array [
                null,
              ],
              "type": "ul",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Method",
                "className": "singleRecipe__sub-title",
              },
              "ref": null,
              "rendered": "Method",
              "type": "h2",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  null,
                ],
                "className": "singleRecipe__list--method",
              },
              "ref": null,
              "rendered": Array [
                null,
              ],
              "type": "ol",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Delete Recipe",
                "className": "singleRecipe__deleteButton",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Delete Recipe",
              "type": "button",
            },
          ],
          "type": "div",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "page": "singleRecipe",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="singleRecipe__img"
            style={
                        Object {
                                    "backgroundImage": "url()",
                                  }
            }
/>,
          <div
            className="singleRecipe__textcontainer--outer"
>
            <div
                        className="singleRecipe__textcontainer--inner"
            >
                        <Link
                                    className="singleRecipe__backarrow"
                                    replace={false}
                                    to="/recipes"
                        >
                                    <i
                                                className="ion-ios-arrow-back"
                                    />
                        </Link>
                        <h1
                                    className="singleRecipe__title"
                        />
                        <h2
                                    className="singleRecipe__sub-title"
                        >
                                    Ingredients
                        </h2>
                        <ul
                                    className="singleRecipe__list--ingredients"
                        >
                                    
                        </ul>
                        <h2
                                    className="singleRecipe__sub-title"
                        >
                                    Method
                        </h2>
                        <ol
                                    className="singleRecipe__list--method"
                        >
                                    
                        </ol>
                        <button
                                    className="singleRecipe__deleteButton"
                                    onClick={[Function]}
                        >
                                    Delete Recipe
                        </button>
            </div>
</div>,
          <Connect(Navbar)
            page="singleRecipe"
/>,
        ],
        "className": "singleRecipe__container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "singleRecipe__img",
            "style": Object {
              "backgroundImage": "url()",
            },
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="singleRecipe__textcontainer--inner"
>
              <Link
                            className="singleRecipe__backarrow"
                            replace={false}
                            to="/recipes"
              >
                            <i
                                          className="ion-ios-arrow-back"
                            />
              </Link>
              <h1
                            className="singleRecipe__title"
              />
              <h2
                            className="singleRecipe__sub-title"
              >
                            Ingredients
              </h2>
              <ul
                            className="singleRecipe__list--ingredients"
              >
                            
              </ul>
              <h2
                            className="singleRecipe__sub-title"
              >
                            Method
              </h2>
              <ol
                            className="singleRecipe__list--method"
              >
                            
              </ol>
              <button
                            className="singleRecipe__deleteButton"
                            onClick={[Function]}
              >
                            Delete Recipe
              </button>
</div>,
            "className": "singleRecipe__textcontainer--outer",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Link
                  className="singleRecipe__backarrow"
                  replace={false}
                  to="/recipes"
>
                  <i
                                    className="ion-ios-arrow-back"
                  />
</Link>,
                <h1
                  className="singleRecipe__title"
/>,
                <h2
                  className="singleRecipe__sub-title"
>
                  Ingredients
</h2>,
                <ul
                  className="singleRecipe__list--ingredients"
>
                  
</ul>,
                <h2
                  className="singleRecipe__sub-title"
>
                  Method
</h2>,
                <ol
                  className="singleRecipe__list--method"
>
                  
</ol>,
                <button
                  className="singleRecipe__deleteButton"
                  onClick={[Function]}
>
                  Delete Recipe
</button>,
              ],
              "className": "singleRecipe__textcontainer--inner",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": <i
                    className="ion-ios-arrow-back"
/>,
                  "className": "singleRecipe__backarrow",
                  "replace": false,
                  "to": "/recipes",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "ion-ios-arrow-back",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "i",
                },
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "",
                  "className": "singleRecipe__title",
                },
                "ref": null,
                "rendered": "",
                "type": "h1",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Ingredients",
                  "className": "singleRecipe__sub-title",
                },
                "ref": null,
                "rendered": "Ingredients",
                "type": "h2",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    null,
                  ],
                  "className": "singleRecipe__list--ingredients",
                },
                "ref": null,
                "rendered": Array [
                  null,
                ],
                "type": "ul",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Method",
                  "className": "singleRecipe__sub-title",
                },
                "ref": null,
                "rendered": "Method",
                "type": "h2",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    null,
                  ],
                  "className": "singleRecipe__list--method",
                },
                "ref": null,
                "rendered": Array [
                  null,
                ],
                "type": "ol",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Delete Recipe",
                  "className": "singleRecipe__deleteButton",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "Delete Recipe",
                "type": "button",
              },
            ],
            "type": "div",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "page": "singleRecipe",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;

exports[`SingleRecipe when \`recipe\` props has not been set yet should only render a loading message <div> 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SingleRecipe
    deleteRecipe={[Function]}
    getSingleRecipe={[Function]}
    match={
        Object {
            "params": Object {
              "id": 2,
            },
          }
    }
    recipe={undefined}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "Loading...",
    },
    "ref": null,
    "rendered": "Loading...",
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Loading...",
      },
      "ref": null,
      "rendered": "Loading...",
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": false,
  },
}
`;
